src/model/implementation/i18n.ts

```diff
@@ -0,0 +1,285 @@
+import * as pluralize from 'pluralize';
+import memorize from 'memorize-decorator';
+import { globalContext } from '../../config/global';
+import { I18N_GENERIC, I18N_WARNING } from '../../meta-schema/constants';
+import { NAMESPACE_SEPARATOR } from '../../schema/constants';
+import {
+    arrayStartsWith, capitalize, compact, decapitalize, flatMap, groupArray, mapFirstDefined, mapValues
+} from '../../utils/utils';
+import { FieldI18nConfig, I18nConfig, NamespaceI18nConfig } from '../config/i18n';
+import { ModelComponent, ValidationContext } from '../validation/validation-context';
+import { Field } from './field';
+import { ObjectTypeBase } from './object-type-base';
+
+export class ModelI18n implements ModelComponent {
+
+    private readonly languageLocalizationProvidersByLanguage: ReadonlyMap<string, LanguageLocalizationProvider>;
+
+    constructor(input: ReadonlyArray<I18nConfig>) {
+        // collect configs by language and create one list of namespaces per language
+        // collect all countries for which namespaces must be created
+        const configsByLanguage = groupArray(input, config => config.language);
+        // const namespacesByCountryPrepared
+        const namespacesByCountry = new Map<string, LanguageLocalizationProvider>();
+        Array.from(configsByLanguage.keys()).forEach(language =>
+            namespacesByCountry.set(language, new LanguageLocalizationProvider(flatMap(configsByLanguage.get(language)!,
+                config => flattenNamespaceConfigs(config.namespaceContent, config.namespacePath)
+                    .map(flatNamespace => new I18nNamespace(flatNamespace)))))
+        );
+        this.languageLocalizationProvidersByLanguage = namespacesByCountry;
+    }
+
+    public validate(context: ValidationContext): void {
+    }
+
+    @memorize()
+    public getTypeLocalization(type: ObjectTypeBase, resolutionOrder: ReadonlyArray<string>): TypeLocalization {
```

Comment:
I consider the `Base` types an implementation detail, the regular union types should be used instead. Also, can be changed to `Type` when `namespacePath` is there (see other comment)